### PR TITLE
Lock To Any [1.3.1]

### PR DESCRIPTION
**Pull request by hayesgm**
This patch adds "Lock to Any" which allows a user to lock an asset destined for either his/her own account or for an account from a different chain, per se. As Eth is the only current chain, this is mostly future proofing.

We also take this time to explore how to perform a safe upgrade. Specifically, the Starport events change after the Starport is upgraded. As such, we will upgrade the chain to a version (1.3.1) that will parse the old and new events from the Starport. Then we will upgrade the Starport. Finally, we will in a future version remove support for the old events.

We actually test that this upgrade pattern works in a new integration test. See `integration/__tests__/upgrade_to_131_scen.js`

Note: this unreverts "Lock to Any" and subsumes #210

### CODE REVIEW

File: integration/__tests__/lock_test.js [hunk 1]
```diff
@@ -1,80 +0,0 @@
-const {
```
Comment (hayesgm) [inline integration/__tests__/lock_test.js:1]: This is very out-dated compared to the new scenarios that it provides little to no value.

File: integration/__tests__/upgrade_to_131_scen.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+const { decodeCall, getEventData } = require('../util/substrate');
+const { bytes32 } = require('../util/util');
+
+let scen_info = {
+  tokens: [
+    { token: "zrx", balances: { ashley: 1000 } }
+  ],
+};
+
+buildScenarios('Upgrade to 1.3.1', scen_info, [
```
Comment (hayesgm) [inline integration/__tests__/upgrade_to_131_scen.js:13]: This test is the crux of showing that upgrades will be safe.

File: integration/util/download.js [hunk 3]
```diff
@@ -19,24 +19,46 @@ function download(url, path, options = {}, handle = null) {
     ...options,
   };
   return new Promise(async (resolve, reject) => {
-    let f = handle || await fs.open(path, 'w');
     let bail = false;
+    let f_;
```
Comment (hayesgm) [inline integration/util/download.js:23]: Here we properly handle the case where we get a 404 (we used to write the 404 message to the file)

File: integration/util/scenario/eth.js [hunk 4]
```diff
@@ -5,13 +5,15 @@ const { readContractsFile, deployContract, getContractAt } = require('../ethereu
 const { genPort } = require('../util');
 
 class Eth {
-  constructor(ethInfo, web3, web3Url, accounts, ganacheServer, ctx) {
+  // TODO: Do the construct
+  constructor(ethInfo, web3, web3Url, accounts, ganacheServer, version, ctx) {
```
Comment (hayesgm) [inline integration/util/scenario/eth.js:9]: Solidity contracts are now versioned so we can deploy older contracts for upgrade and compatibility tests.

File: integration/util/substrate.js [hunk 5]
```diff
@@ -42,6 +43,11 @@ function waitForEvent(api, pallet, method, opts = {}) {
 
   let resolve, reject;
   let promise = new Promise((resolve_, reject_) => {
+    if (opts.timeout) {
```
Comment (hayesgm) [inline integration/util/substrate.js:46]: This timeout should be helpful in stopping "forever running" tests.

File: types.json [hunk 6]
```diff
@@ -12,7 +12,7 @@
   "AssetQuantity": "Quantity",
   "CashQuantity": "Quantity",
   "USDQuantity": "Quantity",
-  "LiquidityFactor": "Factor",
+  "LiquidityFactor": "(u128)",
```
Comment (coburncoburn) [inline types.json:15]: LiquidityFactor in here 2x?

File: types.json [hunk 7]
```diff
@@ -414,6 +414,7 @@
     }
   },
   "Source10": "AccountRS",
+  "LiquidityFactor": "u128",
```
Comment (coburncoburn) [inline types.json:417]: other liquidity factor (u128) or u128?

**Review by coburncoburn — APPROVED** [review]
probable error in types.json

File: integration/util/substrate.js (hunk 5 above)
Comment (waynenilsen) [reply integration/util/substrate.js:46]: yes this helped me a lot thank you

File: pallets/cash/src/internal/events.rs [hunk 8]
```diff
@@ -108,6 +108,13 @@ pub fn receive_event<T: Config>(
     event: ChainLogEvent,
     signature: ValidatorSig,
 ) -> Result<(), Reason> {
+    log!(
```
Comment (waynenilsen) [inline pallets/cash/src/internal/events.rs:111]: Probably too chatty (?) not sure if we want to address this in some broader way though where during sync you don't show all the logs and you start showing them when you catch up.

File: types.json (hunk 6 above)
Comment (waynenilsen) [inline types.json:15]: i don't think this is valid json due to reused key in the same map `LiquidityFactor`

File: pallets/cash/src/internal/lock.rs [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+use crate::{
+    chains::ChainAccount,
+    core::{repay_and_supply_principal, sub_principal_amounts},
+    reason::Reason,
+    types::{CashIndex, CashPrincipalAmount},
+    CashPrincipals, ChainCashPrincipals, Config, Event, GlobalCashIndex, Module,
+    TotalCashPrincipal,
+};
+
+pub fn lock_cash_principal_internal<T: Config>(
```
Comment (waynenilsen) [inline pallets/cash/src/internal/lock.rs:14]: note to other reviewers, this function diff is really something like this afaict 
image

**Review by waynenilsen — APPROVED** [review]
Looks good 👍 

One tiny nit; I think we might want `LockV1` and `LockV2` rather than `LockOld` and `Lock` but that is really just a desire to avoid the potential dreaded `OldOld`.

~I'm not sure it is worth doing this as an integration test (or maybe this is what the upgrade test does) but I think we should "try out" the upgrade process locally with lock events before upgrade and after including sync from 0.~

The point of this PR was to add `integration/__tests__/upgrade_to_131_scen.js` the rest was already approved. From that perspective, I ran the test and read through it. It makes sense to me and seems to prove out the upgrade path.

File: pallets/cash/src/internal/events.rs (hunk 8 above)
Comment (hayesgm) [reply pallets/cash/src/internal/events.rs:111]: Yeah, I think we'll remove many of them once we get things stabilized.

File: types.json (hunk 7 above)
Comment (hayesgm) [reply types.json]: I just reverted these changes, thanks.

**Pull request merged**